src/nodes/chunk_dispatch.c

```diff
@@ -131,7 +131,11 @@ ts_chunk_dispatch_get_chunk_insert_state(ChunkDispatch *dispatch, Point *point,
 	{
 		Chunk *new_chunk;
 
-		new_chunk = ts_hypertable_get_or_create_chunk(dispatch->hypertable, point);
+		new_chunk = ts_hypertable_find_chunk_for_point(dispatch->hypertable, point);
+		if (new_chunk == NULL)
+		{
+			new_chunk = ts_hypertable_create_chunk_for_point(dispatch->hypertable, point);
+		}
```

Comment:
Unfortunately the second check is required, it's done after locking, and before that the chunk can be created by a concurrent transaction.

Reply:
I was actually referring to the first check: is that really needed? There is already a check inside `ts_hypertable_create_chunk_for_point` that will return the chunk if it already exists.

Reply:
The check in `ts_hypertable_create_chunk_for_point` is performed under a `ShareUpdateExclusiveLock` on the hypertable, to serialize concurrent creation of the chunks. We don't want to take this lock unless we have to create the chunk, that's why we first search for the chunk w/o the lock in `ts_hypertable_find_chunk_for_point`. So the first check serves as a "fast path" for the "usual" case when the chunk already exists. This fast path is normally taken for the overwhelming majority of the rows inserted in the chunk -- that is, all except the first one.

Reply:
I added a comment about this.

Reply:
So this means that to use the API correctly it is necessary to first check for the existence of a chunk with `ts_hypertable_find_chunk_for_point` before calling `ts_hypertable_create_chunk_for_point`? Looks like a very cumbersome interface to use and it sounds like this check should be incorporated into the `ts_hypertable_create_chunk_for_point` instead to avoid making mistakes when executing a fetch-or-create.

It would also allow you to avoid the second check since you can upgrade the lock instead of releasing the read lock and re-take a shared update exclusive lock.

However, I note that `get_target_chunk` in `dist_copy.c` actually want to take special action when the chunk does not exist, which would also have to be incorporated into the function instead of putting this requirement on the caller.

However, this is out of scope for the PR, so I leave it to you to decide if you want to do this refactoring.

Reply:
Also note that the "fast path" when the chunk already exists is not affected by the `SHARE UPDATE EXCLUSIVE` lock: it will only block other threads that try to create the chunk at the same time, not reads of the table.

Reply:
Yes, the reason I wanted to keep these functions separate is that some callers need to make a decision based on whether it is a newly created chunk or the old one. Good point about upgrading the lock, I didn't think about that. I think that with #4285, the requirements to this API might change, so I'll revisit it.